packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -121,6 +121,19 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
         }));
     }
 
+    /**
+     * Convenience method for creating a new {@link Stage},
+     * and adding it to this Pipeline.
+     *
+     * @param name the name of the newly created Stage
+     * @returns the newly created Stage
+     */
+    public addStage(name: string): Stage {
+        return new Stage(this.parent!, name, {
```

Comment:
Sorry, I'm not following your example.

This will also fail:

```ts
  constructor(...) {
    super(...);

    new s3.Bucket(this, 'Hello');

    const pl = new codepipeline.Pipeline(this, 'MyPipeline');
    const stage = new codepipeline.Stage(this, 'Hello', { pipeline: pl }); // BOOM! Two children with the same name ("Hello").
  }
```

Which is exactly what we want - `pipeline.addStage('x')` should be exactly equivalent to `new Stage(this, 'x', { pipeline })`.

> Also, since construct ID must be unique, it means no two stages can have the same name. I don't believe this is a constraints in CodePipeline, but it will be a constraint here. 

It is in fact a constraint in CodePipeline (and one that we explicitly validate in the CodePipeline L2).

Reply:
As a CDK developer, I expected the code `new Stage(this, 'Hello')` to fail (since I defined two constructs under the same parent) but not `addStage(`Hello`)` (which hides the fact that a stage is a construct).

Not sure I understand why for some reason you insist on pinning the inner stage construct to the parent's parent?

Reply:
Following up on our chat conversation: I don't believe it is currently priority to make `addStage` and `new Stage` seamlessly interchangeable, and I do think there's value in maintaining the lexical and contextual scope when instantiating the inner `Stage` construct. If we could enforce this pattern by the compiler (and maybe at some point we will), we would have done that. 

Let's just decide that instantiation of constructs __always__ looks like this:

```js
new MyConstruct(this, '<id>', props)
```

Since `<id>` is scoped by `this`, using a different scope (i.e. `this.parent`) can cause conflicts with that other scope, that users can't fix.